### PR TITLE
Rework static filters

### PR DESCRIPTION
**Pull request by nmanu1**
This PR reworks static filters to provide a better developer experience. A new `StaticFilters` model is added that replaces the existing `CombinedFilter`. `Filter` is renamed to `FieldValueFilter` in order to more accurately describe what the object represents. Now static filters can be either a field value filter or a nested object that is composed by combining field value filters. Also, I added Jest tests for some edge cases when serializing static filters.

J=SLAP-2325
TEST=auto, manual

See that the new and existing Jest tests pass. Spin up the test-site and see that requests can be successfully fired when specifying simple or nested static filters and the correct response is received.

### CODE REVIEW

File: src/models/searchservice/request/VerticalSearchRequest.ts [hunk 1]
```diff
@@ -39,7 +38,7 @@ export interface VerticalSearchRequest extends SearchRequest {
   /** Enables session tracking. */
   sessionTrackingEnabled?: boolean,
   /** The static filters to apply to the search. */
-  staticFilters?: CombinedFilter | Filter,
+  staticFilters?: StaticFilter,
```
Comment (tmeyer2115) [inline src/models/searchservice/request/VerticalSearchRequest.ts:41]: This feels a bit odd, the attribute name is plural, but we only allow a singular value.

File: src/models/searchservice/request/StaticFilter.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
+import { FieldValueFilter } from './FieldValueFilter';
+
+/**
+ * Represents a static filter that will be used to refine results.
+ *
+ * @public
+ */
+export type StaticFilter =
+	{ kind: 'fieldValue', value: FieldValueFilter } |
+	{ kind: 'combination', combinator: FilterCombinator, children: StaticFilter[] };
```
Comment (tmeyer2115) [inline src/models/searchservice/request/StaticFilter.ts:10]: Are these indentations correct?

File: src/models/searchservice/response/DisplayableFacet.ts [hunk 3]
```diff
@@ -1,15 +1,11 @@
...
-import { NumberRangeValue } from '../common/NumberRangeValue';
 
 /**
  * A {@link Facet} which contains extra fields meant to be displayed to the end user.
  *
  * @public
  */
 export interface DisplayableFacet extends Facet {
-  /** {@inheritDoc Facet.fieldId} */
-  fieldId: string,
```
Comment (tmeyer2115) [inline src/models/searchservice/response/DisplayableFacet.ts:12]: Is this now in `Facet`?

File: src/models/searchservice/response/DisplayableFacet.ts [hunk 4]
```diff
@@ -22,10 +18,6 @@ export interface DisplayableFacet extends Facet {
  * @public
  */
 export interface DisplayableFacetOption extends FacetOption {
-  /** {@inheritDoc Matcher} */
-  matcher: Matcher,
-  /** {@inheritDoc FacetOption.value} */
-  value: string | number | boolean | NumberRangeValue,
```
Comment (tmeyer2115) [inline src/models/searchservice/response/DisplayableFacet.ts:28]: These are now in `FacetOption`, right?

File: src/serializers/serializeStaticFilters.ts [hunk 5]
```diff
@@ -1,32 +1,41 @@
...
-import { Filter } from '../models/searchservice/request/Filter';
+import { FilterCombinator, StaticFilter } from '../models/searchservice/request/StaticFilter';
+import { FieldValueFilter } from '../models/searchservice/request/FieldValueFilter';
 import { ApiStaticFilters, ApiFilter } from '../models/searchservice/request/ApiStaticFilters';
 
 export function serializeStaticFilters(
-  filter: CombinedFilter | Filter): string | undefined {
-  if (isCombinedFilter(filter)) {
-    return JSON.stringify(shapeCombinedFilterForApi(filter));
+  filter: StaticFilter): string | undefined {
```
Comment (tmeyer2115) [inline src/serializers/serializeStaticFilters.ts:7]: Can we move this to the previous line?

File: src/serializers/serializeStaticFilters.ts [hunk 6]
```diff
@@ -1,32 +1,41 @@
...
 
-  return JSON.stringify(shapeFilterForApi(filter));
+  return JSON.stringify(shapeFieldValueFilterForApi(filter.value));
 }
 
-function shapeCombinedFilterForApi(combinedFilter: CombinedFilter): ApiStaticFilters {
+function shapeCombinedFilterForApi(
+  combinator: FilterCombinator,
+  filters: StaticFilter[]
+): ApiStaticFilters | undefined {
```
Comment (tmeyer2115) [inline src/serializers/serializeStaticFilters.ts:19]: Can the method declaration fit on one line?

File: src/models/searchservice/response/DisplayableFacet.ts (hunk 3 above)
Comment (nmanu1) [reply src/models/searchservice/response/DisplayableFacet.ts:12]: yes, this property was redundant even before the rework because it's the same as the property in `Facet`. there was no reason to specify it again after extending `Facet`

File: src/models/searchservice/response/DisplayableFacet.ts (hunk 4 above)
Comment (nmanu1) [reply src/models/searchservice/response/DisplayableFacet.ts:28]: this was the same thing as for `DisplayableFacet` where some properties were being redefined after extending the base model even though nothing was being changed. so I just removed the redundancy

File: src/serializers/serializeStaticFilters.ts (hunk 6 above)
Comment (nmanu1) [reply src/serializers/serializeStaticFilters.ts:19]: no, eslint complains about the line being too long if it's all on the same one

File: src/models/searchservice/request/StaticFilter.ts (hunk 2 above)
Comment (nmanu1) [reply src/models/searchservice/request/StaticFilter.ts]: updated!

File: src/serializers/serializeStaticFilters.ts (hunk 5 above)
Comment (nmanu1) [reply src/serializers/serializeStaticFilters.ts]: updated!

**Comment by oshi97** [comment]
Are the main changes
1. renaming
2. a `kind` discriminator for filters, either `fieldValue` or `combination`?

**Comment by nmanu1** [comment]
> Are the main changes
> 
> 1. a `kind` discriminator for filters, either `fieldValue` or `combination`
> 2. renaming associated with the above
> 
> ?

yes, the main things we wanted were
1. a single interface for static filters instead of having it explicitly split up into 2 types based on how we represent them for the API
2. the `kind` discriminator to make type checking easier for consumers of the Headless static filters state
3. renaming `Filter` to `FieldValueFilter` to make it more clear, because facets and static filters are filters too
4. and then the renaming associated with the above changes

File: src/models/searchservice/request/VerticalSearchRequest.ts (hunk 1 above)
Comment (nmanu1) [reply src/models/searchservice/request/VerticalSearchRequest.ts:41]: I agree, this looks a little weird. Although, this was the case before also where we accepted a single value (either a `Filter` or a `CombinedFilter`) for the plural `staticFilters`. It was just slightly obscured because of the union type.
Would it make sense to rename the attribute to `staticFilter`? Because that's what the user is giving us; it's just that they can combine a bunch of field value filters together if they want to in their one static filter

Comment (tmeyer2115) [reply src/models/searchservice/request/VerticalSearchRequest.ts:41]: Yeah, renaming to static filter sounds good!

**Review by tmeyer2115 — APPROVED** [review]
